### PR TITLE
Added capture to keys

### LINKED ISSUE #143: Capture keystroke
If user clicks on the key with a middle mouse button, capture keystroke is activated. If the user clicks outside of the current active key, capture keystroke is canceled. When capture keystroke is active key is blinking.

### PR DESCRIPTION
**Pull request by NejcZdovc**
Resolves #143

### CODE REVIEW

**Comment by mondalaci** [comment]
This seems to work perfectly, and I like the pulsing animation very much.

How about using red instead of gray background color for the key? That'd be way more noticable and would resemble the red color of recording buttons, giving a hint to the user. At the same time, it'd be nice to display a record icon and "Capturing..." text below the icon inside of the key.

**Comment by NejcZdovc** [comment]
I wouldn't go with a text, because rectangles are small and if we would introduce other languages we could have problems. Will try to do it with a red color and a circle for the icon

**Comment by mondalaci** [comment]
That should also cut it :+1:

**Comment by mondalaci** [comment]
The latest version is the bee's knees. I'm happy with it.

File: src/components/keyboard/slider/keyboard-slider.component.ts [hunk 1]
```diff
@@ -81,6 +81,7 @@ export class KeyboardSliderComponent implements OnChanges {
     @Input() keybindAnimationEnabled: boolean;
     @Output() keyClick = new EventEmitter();
     @Output() keyHover = new EventEmitter();
+    @Output() captured = new EventEmitter();
```
Comment (fjozsef) [inline src/components/keyboard/slider/keyboard-slider.component.ts]: Rename to `capture`

File: src/components/svg/keyboard/svg-keyboard.component.ts [hunk 2]
```diff
@@ -13,6 +13,7 @@ export class SvgKeyboardComponent implements OnInit {
     @Input() keybindAnimationEnabled: boolean;
     @Output() keyClick = new EventEmitter();
     @Output() keyHover = new EventEmitter();
+    @Output() captured = new EventEmitter();
```
Comment (fjozsef) [inline src/components/svg/keyboard/svg-keyboard.component.ts]: Rename to `capture`

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts [hunk 3]
```diff
@@ -61,45 +71,122 @@ export class SvgKeyboardKeyComponent implements OnInit, OnChanges, OnDestroy {
     @Input() keyAction: KeyAction;
     @Input() keybindAnimationEnabled: boolean;
     @Output() keyClick = new EventEmitter();
+    @Output() captured = new EventEmitter();
```
Comment (fjozsef) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Rename to `capture`

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts [hunk 4]
```diff
@@ -61,45 +71,122 @@ export class SvgKeyboardKeyComponent implements OnInit, OnChanges, OnDestroy {
...
 
+    public changeAnimation: string = 'inactive';
+    public recordAnimation: string;
+
     private labelSource: any;
     private labelType: LabelTypes;
     private macros: Macro[];
     private subscription: Subscription;
-    private animation: string = 'inactive';
+    private record: boolean;
```
Comment (fjozsef) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Rename to `recording`

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts [hunk 5]
```diff
@@ -61,45 +71,122 @@ export class SvgKeyboardKeyComponent implements OnInit, OnChanges, OnDestroy {
...
-        /* tslint:enable:no-string-literal */
     }
 
     ngOnDestroy() {
         this.subscription.unsubscribe();
     }
 
-    animationDone() {
-        this.animation = 'inactive';
+    changeDone() {
```
Comment (fjozsef) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Rename to `onChangeAnimationDone` or `onChangeDone` depending which you prefer.

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts [hunk 6]
```diff
@@ -61,45 +71,122 @@ export class SvgKeyboardKeyComponent implements OnInit, OnChanges, OnDestroy {
...
+        this.changeAnimation = 'inactive';
+    }
+
+    recordingDone() {
```
Comment (fjozsef) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Rename to `onRecordingAnimationDone` or `onRecordingDone` depending which you prefer.

File: src/components/svg/module/svg-module.component.ts [hunk 7]
```diff
@@ -16,6 +16,7 @@ export class SvgModuleComponent {
     @Input() keybindAnimationEnabled: boolean;
     @Output() keyClick = new EventEmitter();
     @Output() keyHover = new EventEmitter();
+    @Output() captured = new EventEmitter();
```
Comment (fjozsef) [inline src/components/svg/module/svg-module.component.ts]: Rename to `capture`

File: src/components/svg/wrap/svg-keyboard-wrap.component.ts [hunk 8]
```diff
@@ -267,4 +288,11 @@ export class SvgKeyboardWrapComponent implements OnInit, OnChanges {
         this.currentLayer = index;
     }
 
+    private modifierMapper(x: number) {
```
Comment (fjozsef) [inline src/components/svg/wrap/svg-keyboard-wrap.component.ts]: Don't duplicate code, rather move this function to the `MapperService`

File: src/components/svg/module/svg-module.component.ts (hunk 7 above)
Comment (NejcZdovc) [inline src/components/svg/module/svg-module.component.ts]: Done

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts (hunk 6 above)
Comment (NejcZdovc) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Done

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts (hunk 5 above)
Comment (NejcZdovc) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Done

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts (hunk 4 above)
Comment (NejcZdovc) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Done

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts (hunk 3 above)
Comment (NejcZdovc) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Done

File: src/components/svg/keyboard/svg-keyboard.component.ts (hunk 2 above)
Comment (NejcZdovc) [inline src/components/svg/keyboard/svg-keyboard.component.ts]: Done

File: src/components/keyboard/slider/keyboard-slider.component.ts (hunk 1 above)
Comment (NejcZdovc) [inline src/components/keyboard/slider/keyboard-slider.component.ts]: Done

File: src/components/svg/wrap/svg-keyboard-wrap.component.ts (hunk 8 above)
Comment (NejcZdovc) [inline src/components/svg/wrap/svg-keyboard-wrap.component.ts]: Moved to the mapper service

**Comment by NejcZdovc** [comment]
I tried to trigger multiple recording, but couldn't. How do you trigger multiple recordings?

Default action for key down and mouse up is prevented. Why should we block mouse scroll when recording? We are recording key strokes, not mouse mouse related actions.

**Comment by fjozsef** [comment]
Simply clicking with middle mouse button on multiple keys.

The middle click should be prevented, which starts the recording.

File: src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts [hunk 9]
```diff
@@ -61,45 +71,130 @@ export class SvgKeyboardKeyComponent implements OnInit, OnChanges, OnDestroy {
...
     }
 
-    animationDone() {
-        this.animation = 'inactive';
+    onChangeAnimationDone() {
+        this.changeAnimation = 'inactive';
+    }
+
+    onRecordingAnimationDone() {
+        if (this.recording) {
```
Comment (fjozsef) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: How about changing this to `if (this.recording && this.recordAnimation === 'inactive')` and the next line could be assigning `active` ?

Comment (NejcZdovc) [inline src/components/svg/keys/svg-keyboard-key/svg-keyboard-key.component.ts]: Correct, this was a leftover of the old implementation.

**Pull request merged**